pkg/driver_cryptocell_310/psa_cryptocell_310/cipher_chacha20.c

```diff
@@ -0,0 +1,135 @@
+        !CHECK_POINTER_DMA_ACCESS(input) ||
+        !CHECK_POINTER_DMA_ACCESS(output)) {
+        return PSA_ERROR_DATA_INVALID;
+    }
+
+    if (output_size < (input_length + CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES)) {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    if (key_buffer_size != CRYS_CHACHA_KEY_MAX_SIZE_IN_BYTES) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    uint8_t *nonce = output;
+    uint8_t *data_out = &output[CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES];
+    status = psa_generate_random(nonce, CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    cryptocell_310_enable();
+    CRYSError_t periph_status = CRYS_CHACHA(nonce, CRYS_CHACHA_Nonce96BitSize,
+                                            key_buffer, 0UL,
+                                            CRYS_CHACHA_Encrypt,
+                                            (uint8_t *) input,
+                                            input_length,
+                                            data_out);
+    cryptocell_310_disable();
+    status = CRYS_to_psa_error(periph_status);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+
+    *output_length = input_length + CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES;
+    return PSA_SUCCESS;
+}
+
+psa_status_t psa_cipher_chacha20_decrypt(uint8_t *key_buffer,
+                                         size_t key_buffer_size,
+                                         const uint8_t *input,
+                                         size_t input_length,
+                                         uint8_t *output,
+                                         size_t output_size,
+                                         size_t *output_length)
+{
+    DEBUG("Peripheral ChaCha20 Cipher decryption");
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if (!CHECK_POINTER_DMA_ACCESS(key_buffer) ||
+        !CHECK_POINTER_DMA_ACCESS(input) ||
+        !CHECK_POINTER_DMA_ACCESS(output)) {
+        return PSA_ERROR_DATA_INVALID;
+    }
+
+    if ((key_buffer_size != CRYS_CHACHA_KEY_MAX_SIZE_IN_BYTES) ||
+        (input_length < CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES)) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (output_size < (input_length - CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES)) {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    }
+
+    uint8_t *nonce = (uint8_t *) input;
+    uint8_t *data_in = (uint8_t *) &input[CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES];
+    size_t data_size = input_length - CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES;
+
+    cryptocell_310_enable();
+    CRYSError_t periph_status = CRYS_CHACHA(nonce, CRYS_CHACHA_Nonce96BitSize,
+                                            key_buffer, 0UL,
+                                            CRYS_CHACHA_Decrypt,
+                                            data_in,
+                                            data_size,
+                                            output);
```

Comment:
```suggestion
    const uint8_t *nonce = input;
    const uint8_t *data_in = &input[CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES];
    size_t data_size = input_length - CRYS_CHACHA_NONCE_MAX_SIZE_IN_BYTES;

    cryptocell_310_enable();
    CRYSError_t periph_status = CRYS_CHACHA((uint8_t *)nonce, CRYS_CHACHA_Nonce96BitSize,
                                            key_buffer, 0UL,
                                            CRYS_CHACHA_Decrypt,
                                            (uint8_t *)data_in,
                                            data_size,
                                            output);
```

to make it obvious that we cast only because of the CRYS_CHACHA declaration, not because we actually want to be able to change the buffer content.